### PR TITLE
LPS-135604 Expanding calculate rule to complex expressions

### PR DESCRIPTION
**Pull request by rijema**
Related Issue: https://issues.liferay.com/browse/LPS-135604

Hi! I'm opening this PR to include the fix for LPS-135604 reported in the QA evaluation. This pull request complements the #843 . I'm available in case of any doubts. 

Thanks, team!

### CODE REVIEW

**Comment by liferay-forms** [comment]
Hi, you don’t need to run the CI commands anymore, I’ll do it for you and let you know if any issue happens.

**Comment by liferay-forms** [comment]
ci:test:sf

**Comment by liferay-forms** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorRuleHelper.java [hunk 1]
```diff
@@ -151,8 +158,42 @@ protected boolean containsAction(
 			action -> Objects.equals(setPropertyAction, action));
 	}
 
+	protected boolean containsCalculatedAction(
```
Comment (rijema) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorRuleHelper.java:161]: In this function, I'm evaluating more complex expressions affected by a calculate rule. The purpose here is, for each rule, verify if the field can get the correct value after gets out of some action. As the field only gets here if they are not affected by the rule anymore and the _defaultValue_ is the  previous value defined by the rule, I'm checking if the current value is the same as if I had to evaluate the field. The call for the method calculateAction returns all the fields evaluated by a specific rule, this was necessary for compare both values. The conditional in line 171 was created to minimize the number of time that fields goes into this calls and do the evaluation. In the end (182), there are two verifications: one if the values are the same as I said before, but also if the result of the fields are 0 (the use case I tried was when fields are empty, the affected field should return to the predefined or empty value).

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorHelper.java [hunk 2]
```diff
@@ -172,6 +172,32 @@ public DDMFormEvaluatorEvaluateResponse evaluate() {
 		return formEvaluatorEvaluateResponse.build();
 	}
 
+	protected Map<String, Object> calculateAction(DDMFormRule ddmFormRule) {
```
Comment (rijema) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/helper/DDMFormEvaluatorHelper.java:175]: Here I'm returning the fields values evaluated by some rule. To do that, I had to use the function that literally calculates the fields. And as this function updates the __ddmFormFieldsPropertyChanges_, I'm cleaning the object after invokes this evaluation. Before that, I'm running through the object to populate a auxiliar object that I'm returning in this function to the _containsCalculatedAction_ get all the necessary info.

**Comment by liferay-forms** [comment]
Hey @rijema, can you please check if these failures are related to your changes? Thanks

**Comment by rijema** [comment]
They are related.

**Comment by rijema** [comment]
ci:test:sf

**Comment by rijema** [comment]
ci:test:relevant

**Comment by liferay-forms** [comment]
ci:test:relevant

**Comment by liferay-forms** [comment]
Hey @rijema, can you please check if these failures are related to your changes? Thanks

**Comment by rijema** [comment]
They aren't related.

**Comment by natocesarrego** [comment]
Thanks for your pull, @rijema!

Can we have a call to discuss about it by next Monday on morning? If so, please let me know through Slack.

Thanks

**Comment by rijema** [comment]
Sure @natocesarrego! Just sent a message to you.

**Review by natocesarrego — CHANGES_REQUESTED** [review]

**Comment by natocesarrego** [comment]
I've discussed with @rijema about this pull in a call and I've suggested him a different manner to tackle this problem. Therefore, I'm closing this pull and I'll wait a new one according to my recommendation.

Thanks for your hard work on this issue, @rijema! :slightly_smiling_face:

**Pull request closed without merging**